go/vt/topo/tablet_test.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topo_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/topo/memorytopo"
+)
+
+func TestServerGetTabletsByCell(t *testing.T) {
+	tests := []struct {
+		name      string
+		tablets   int
+		opt       *topo.GetTabletsByCellOptions
+		listError error
+	}{
+		{
+			name:    "negative concurrency",
+			tablets: 1,
+			// Ensure this doesn't panic.
+			opt: &topo.GetTabletsByCellOptions{Concurrency: -1},
+		},
+		{
+			name:    "unsharded",
+			tablets: 1,
+			// Make sure the defaults apply as expected.
+			opt: nil,
+		},
+		{
+			name:    "sharded",
+			tablets: 32,
+			opt:     &topo.GetTabletsByCellOptions{Concurrency: 8},
+		},
+		{
+			name:      "sharded with list error",
+			tablets:   32,
+			opt:       &topo.GetTabletsByCellOptions{Concurrency: 8},
+			listError: topo.NewError(topo.ResourceExhausted, ""),
```

Comment:
I added a temporary log line (which I removed before commit) and verified that we in fact call `GetTablet` on every tablet in the presence of this error.